[Bug fix] remove `mujoco-py` import error for v4+ MuJoCo environments

MischaPanch opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #896 

## Type of change

Please delete options that are not relevant.

- [ ] Documentation only change (no code changed)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Screenshots

Please attach before and after screenshots of the change if applicable.

<!--
Example:

| Before | After |
| ------ | ----- |
| _gif/png before_ | _gif/png after_ |


To upload images to a PR -- simply drag and drop an image while in edit mode and it should upload the image directly. You can then paste that source into the above before/after sections.
-->

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

<!--
As you go through the checklist above, you can mark something as done by putting an x character in it

For example,
- [x] I have done this task
- [ ] I have not done this task
-->

Kallinteris-Andreas commented on gymnasium/envs/mujoco/__init__.py:
```diff
@@ -1,2 +1 @@
-from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv
```
export `MujocoEnv` & `MuJocoPyEnv`

MischaPanch replied:
That counteracts the purpose of the PR - the modules shouldn't be imported in the same context

MischaPanch replied:
It works if old and new mujoco are both installed but it defeats the purpose, the issue is then but resolved. You need to be able to import from mujoco_env without triggering the import error from mujoco_py_env

MischaPanch replied:
I'll merge the files as discussed below, but the import structure still needs to stay as it's now on this branch

Kallinteris-Andreas replied:
Can we at least keep `from gymnasium.envs.mujoco.mujoco_env import MujocoEnv` otherwise many other projects that depend on `MujocoEnv` will have to be updated

MischaPanch replied:
Done

Kallinteris-Andreas commented on gymnasium/envs/mujoco/mujoco_env.py:
```diff
@@ -1,355 +1,21 @@
-        Optionally implement this method, if you need to tinker with camera position and so forth.
-        """
-        raise NotImplementedError
+    raise error.DependencyNotInstalled(
```
Keep the old message
```py
             raise error.DependencyNotInstalled(
                f"{MUJOCO_IMPORT_ERROR}. "
                "(HINT: you need to install mujoco, run `pip install gymnasium[mujoco]`.)"
            )
```

MischaPanch replied:
I adjusted it to contain the HINT: ... (although I personally find it less appropriate, people can install gymnasium or mujoco in very different ways).

Not that I changed the error raising mechanism to `raise from` so the original import error message is also displayed. `raise from` is the recommended approach for such situations, I believe

Kallinteris-Andreas commented on gymnasium/envs/mujoco/mujoco_env_base.py:
```diff
@@ -0,0 +1,420 @@
+        return np.concatenate([self.data.qpos.flat, self.data.qvel.flat])
+
+
+class MuJocoPyEnv(BaseMujocoEnv):
```
Why is this here? (it is also in `mujoco_py_env.py`)

MischaPanch replied:
Was an accident, deleted it

Kallinteris-Andreas reviewed (changes requested):
As I said in the issue with regard to `baseMujocoEnv`, do not move it from `mujoco.py` instead create copy in `mujoco_py_env.py`

MischaPanch commented:
Your pre-commit pipeline was not running on my machine because of some pkg_resources related error. Unfortunately, there was no setup instructions in the contributing.md. Could you either reformat the code yourself or share the development setup with me?

Kallinteris-Andreas reviewed (changes requested):
Sorry for the miscommunication on my part, (since my initial comment in your issue and now, we have marked `mujoco-py` environments as explicitly deprecated)

Here is are the updated requirements happen:
files:
- `mujoco_env.py`:  keep `baseMujocoEnv` & `MujocoEnv`
- `mujoco_py_env.py`: place a copy of `baseMujocoEnv` (and rename it to `baseMujocoPyEnv`) & `MujocoPyEnv`

motivation: `MujocoPyEnv` should be in its own file with nothing external affecting it, `BaseMujocoEnv` will eventually be "merged" with `MujocoEnv`

MischaPanch commented:
I see, makes sense. I'll adjust accordingly and write a block comment in the reasons for the voice duplication

Kallinteris-Andreas commented on gymnasium/envs/mujoco/ant_v5.py:
```diff
@@ -5,7 +5,7 @@
 import numpy as np
 
 from gymnasium import utils
-from gymnasium.envs.mujoco import MujocoEnv
```
no need to change this

MischaPanch replied:
Done

Kallinteris-Andreas commented on tests/envs/mujoco/test_mujoco_v5.py:
```diff
@@ -6,7 +6,8 @@
 import gymnasium as gym
-from gymnasium.envs.mujoco.mujoco_env import BaseMujocoEnv, MujocoEnv
+from gymnasium.envs.mujoco.mujoco_env import MujocoEnv
+from gymnasium.envs.mujoco.mujoco_env_base import BaseMujocoEnv
```
revert

Kallinteris-Andreas reviewed (changes requested):
Hey, sorry for being late review, I missed the request for review


mostly looks good

MischaPanch commented:
Done. One more import in a test file was adjusted to be consistent with the remaining imports (I did a search-replace)

Kallinteris-Andreas commented on gymnasium/envs/mujoco/__init__.py:
```diff
@@ -1,2 +1,2 @@
-from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv
```
We do not want to export `BaseMujocoEnv`

MischaPanch replied:
Done

Kallinteris-Andreas commented on tests/envs/mujoco/test_mujoco_v5.py:
```diff
@@ -6,7 +6,7 @@
 import pytest
 
 import gymnasium as gym
-from gymnasium.envs.mujoco.mujoco_env import BaseMujocoEnv, MujocoEnv
```
I suppose this also needs to be reverted

Kallinteris-Andreas commented on gymnasium/envs/mujoco/mujoco_py_env.py:
```diff
@@ -0,0 +1,342 @@
+
+
+# noinspection DuplicatedCode
+class BaseMujocoEnv(gym.Env[NDArray[np.float64], NDArray[np.float32]]):
```
rename to `BaseMujocoPyEnv`

MischaPanch replied:
Done

Kallinteris-Andreas reviewed (changes requested):
This is looking almost done
To fix the failing test
change
```py
assert isinstance(env, BaseMujocoEnv)
```
to 
```py
assert isinstance(env, BaseMujocoEnv) or isinstance(env, BaseMujocoPyEnv)
```
then rebase to main, and we should be ready for merge

MischaPanch commented:
@Kallinteris-Andreas 

1. I adjusted the `isinstance` (using tuple instead of multiple checks)
2. Adjusted imports as you asked
3. Rebasing resulted in conflicts, so I merged main instead
